Suggestions for Abstract and Introduction (#56)
DanNixon/sine2020_software_guidelines_paper: guidelines_for_collaborative_development_of_sustainable_data_treatment_software.tex

\address[E]{\orgname{Institut Laue-Langevin},\cny{France}\printead[presep={\\}]{e8}}

\begin{abstract}
Scientific software is an essential part of the modern researcher's toolkit.

> Due to large volumes of data and complex analysis techniques, such a software is often expensive to develop and maintain.

Reviewer: ```suggestion
Due to the large volumes of data and the complex analysis techniques required, it is often expensive to develop and maintain scientific software.
```

Therefore, it is imperative to ensure that the software will continue to be supported in the future, even if the current work force leaves the project.
Some of the requirements for sustainable software are for the code base to have a consistent quality and good documentation, to prevent a loss of knowledge.
This paper presents two sets of guidelines. 

> The first is a generic set that could be applied to software projects in a broad range of domains.

Reviewer: Im not sure on the word domain. I think its unclear in this context what that could mean.

Reviewer 2: I think it does make sense (referring to category of problem) but there may be a better word for it here.

Author: I still believe domain is the best word here. I think in this context it is exactly that word that is used most commonly.

Reviewer: I just think fitting domain... too much time do fitting in Mantid

\section{Introduction}
\label{Introduction}

It has become vital for modern neutron and muon sources to ensure that adequate user software is in existence not only for data acquisition but also for data treatment, in order to ensure the production of scientific result rather than just data.

> The reason for this is at least two-fold: first, the demography of the users is changing from scientists building a career in neutron scattering or muon spectroscopy to scientists that consider neutron scattering and muon spectroscopy only as tools among a broader set of tools; they are thus often not knowledgeable about the details behind the data treatment.

Reviewer: should it be demographic instead of demography? 

I dont the the only is needed and I think we should say 
```
career in neutron scattering or muon spectroscopy to scientists that consider it as one of many tools available to researchers. Hence, they are often not knowledgeable about the details behind the neutron and muon data treatment.

Moreover, the users may not have the technical skills in how to analyse their data with programs like {\sc Matlab}, {\sc Python}, {\sc IDL}, etc. 
The second reason is that the advances in science are demanding increasingly greater amounts of data that cannot be analysed by hand.
For example, some instruments at the
European Spallation Source (ESS) are estimated to have data rates beyond ...(\todo{TODO: Tobias paper}) and at modern photon sources the situation is even more severe.
This can be compared to the data rates from ten years ago, where ...(\todo{TODO: Historic example}).

The growing complexity of both data reduction and analysis means that the development of scientific software requires increasingly more resources.

> The consequence of these aspects is that users will inherently depend on software developed by other developers, typically at the hosting facilities or in the user community.

Reviewer: ```suggestion
The consequence of these aspects is that users will inherently depend on software created by other developers, typically at the hosting facilities or in the user community.
```

> This adds significant requirements in terms of quality, trustability, reliability, reproducibility and usability compared to when a user develops software for his own use.

Reviewer: ```suggestion
This adds significant requirements in terms of quality, trustability, reliability, reproducibility and usability compared to when a user develops software for their own use.
```

The lack of guidelines can lead to the code being difficult to read, because the style will change depending on who wrote it.

> In the worst case this may lead to people believing that the code is incorrect, when it is correct, or other way around.

Reviewer: ```suggestion
In the worst case this may lead to people believing that the code is incorrect, when it is correct and vica-versa.
```

The other potential pitfall with this isolated approach of software development is that bugs are less likely to be found, due to less users and less developers.
Also the software may become too cumbersome to develop leading to the project having to start from scratch.
These factors make it more difficult to share software and for other research groups to adapt the software for their own purpose.

Guidelines for software development ensure that the code is of a high quality and that the project progresses effectively.

> A good set of guidelines will help the software be sustainable; that is, reliable, inter-operable, extensible, intuitive, maintainable.

Reviewer: I liked this better as a bullet point list as it was easier to read

> That way it will be easier to continue growing the code base to include new scientific analysis and techniques and add new functionality.

Reviewer: ```suggestion
That way it will be easier to continue growing the code base to include new functionality to the current scientific analysis techniques and to expand into new areas of analysis/research.
```

It would also reduce the effort required to adapt the software for use at other facilities.

Generic software development guidelines (\todo{TODO: cite}) have been common practice for many years.
Specific software guidelines have been created for several scientific domains (\todo{TODO: cite}).
The specific guidelines presented in Section~\ref{SINE2020} were developed as part of SINE2020.
The software projects encompassed in the frame of SINE2020 had a broad range of team sizes: from a couple of developers working very closely on a single piece of software to wide international collaborations.

Although the specific guidelines are aimed at the neutron and muon communities, the generic guidelines reflect what is common practice in software development.
Hence, when it is appropriate the idea of the guideline will be abstracted, so that an alternative can be chosen in the future.

> For instance, {\sc Python} is today the \textit{de facto} standard scripting language in scientific computing even though the first scientific python conference to the best of our knowledge was held as late as in 2002.

Reviewer: ```suggestion
For instance, at the time of writing the \textit{de facto} standard scripting language in scientific computing is , {\sc Python}, even though the first scientific python conference to the best of our knowledge was held as late as in 2002.
```

> Abstracting guidelines in this manner allows to remain relevant as technology evolves.

Reviewer: ```suggestion
Abstracting guidelines in this manner allows them to remain relevant as technology evolves.
```

\item{intuitive}
\item{maintainable.}
\end{itemize}
That way it will be easier to continue growing the code base to include new functionality to the current scientific analysis techniques and to expand into new areas of analysis/research.
It would also reduce the effort required to adapt the software for use at other facilities.

Generic software development guidelines (\todo{TODO: cite}) have been common practice for many years.
% cite IEEE Software, Vol. 24 Number 1 IEEE, January 2007 or similar
Specific software guidelines have been created for several scientific domains (\todo{TODO: cite}).
The specific guidelines presented in Section~\ref{SINE2020} were developed as part of SINE2020.
The software projects encompassed in the frame of SINE2020 had a broad range of team sizes: from a couple of developers working very closely on a single piece of software to wide international collaborations.

Although the specific guidelines are aimed at the neutron and muon communities, the generic guidelines reflect what is common practice in software development.
Hence, when it is appropriate the idea of the guideline will be abstracted, so that an alternative can be chosen in the future.

> For instance, at the time of writing the \textit{de facto} standard scripting language in many fields of scientific computing is {\sc Python}, even though the first scientific python conference to the best of our knowledge was held as late as in 2002.

Reviewer 2: I think this carries the same meaning as what was already there. Unless we can confidently say this then we shouldn't say it.

\section{Introduction}
\label{Introduction}

It has become vital for modern neutron and muon sources to ensure that adequate user software is in existence not only for data acquisition but also for data treatment, in order to ensure the production of scientific result rather than just data.
The reason for this is at least two-fold: first, the demographic of the users is changing from scientists building a career in neutron scattering or muon spectroscopy to scientists that consider them as one of many tools. 
Hence, they are often not knowledgeable about the details behind the neutron and muon data treatment.

> Moreover, the users may not have the technical skills in how to analyse their data with programs like {\sc Matlab}, {\sc Python}, {\sc IDL}, etc. 

Reviewer 2: Do we want to say Python here since we go on to acknowledge its popularity and hint at its simplicity?

Reviewer: I think this is alluding to GUI's (e.g. muon analysis)

Author: I think here we can even omit the list of languages.

The second reason is that the advances in science are demanding increasingly greater amounts of data that cannot be analysed by hand.
For example, some instruments at the

> European Spallation Source (ESS) are estimated to have data rates beyond (\todo{TODO: Tobias paper}) and at modern photon sources the situation is even more severe (\todo{TODO: Mention ESRF upgrade}).

Reviewer 2: @AnthonyLim23 I can't remember exactly but I thought we weren't going to mention photon sources as it is not overly relevant to the audience?

Reviewer: that is a good point

Author: Will remove the mention about photon sources.

This can be compared to the data rates from ten years ago, where (\todo{TODO: Historic example}).

The growing complexity of both data reduction and analysis means that the development of scientific software requires increasingly more resources.
The consequence of these aspects is that users will inherently depend on software created by other developers, typically at the hosting facilities or in the user community.
This adds significant requirements in terms of quality, trustability, reliability, reproducibility and usability compared to when users develop software for their own use.
The lack of guidelines can lead to the code being difficult to read, because the style will change depending on who wrote it.
In the worst case this may lead to people believing that the code is incorrect, when it is correct and vica-versa.

> The other potential pitfall with this isolated approach of software development is that bugs are less likely to be found, due to less users and less developers.

Reviewer 2: I actually think the more developers you have the more issues you will see, rather than a small amount of developers that understand the problem and it's scope very well.

Reviewer: I think it was supposed to mean that with less developers, there is less testing and less people to fix bugs

Reviewer 2: Then it might be too ambiguous, as less developers also means less code changes and less misunderstandings so less bugs.

Reviewer: probably. We are at a chicken and egg like problem.

Author: I will remove the `less developers` clause, since indeed it creates some confusion.

\item{intuitive}
\item{maintainable.}
\end{itemize}
That way it will be easier to continue growing the code base to include new functionality to the current scientific analysis techniques and to expand into new areas of analysis/research.
It would also reduce the effort required to adapt the software for use at other facilities.

Generic software development guidelines (\todo{TODO: cite}) have been common practice for many years.
% cite IEEE Software, Vol. 24 Number 1 IEEE, January 2007 or similar
Specific software guidelines have been created for several scientific domains (\todo{TODO: cite}).
The specific guidelines presented in Section~\ref{SINE2020} were developed as part of SINE2020.
The software projects encompassed in the frame of SINE2020 had a broad range of team sizes: from a couple of developers working very closely on a single piece of software to wide international collaborations.

Although the specific guidelines are aimed at the neutron and muon communities, the generic guidelines reflect what is common practice in software development.
Hence, when it is appropriate the idea of the guideline will be abstracted, so that an alternative can be chosen in the future.

> For instance, at the time of writing the \textit{de facto} standard scripting language in scientific computing is {\sc Python}, even though the first scientific python conference to the best of our knowledge was held as late as in 2002.

Reviewer 2: Are we certain of this enough to claim it here? Maybe we can say neutron and muon analysis, but I'm not confident we can say all of scientific computing uses Python as the go to language.

(I realise you didn't actually write this, I only just thought this after rereading this after some time)

Reviewer: yes we can do.

Reviewer 2: Then we need to specify we are talking about particle physics, a lot of other fields will use something other than Python.
